docs/compilers/terrapin.md

```diff
@@ -0,0 +1,125 @@
+- Validate binaries produced on Linux / Mac: the target of this milestone is NuPkgs produced on Linux / Mac. This will require Terrapin to involve multiple machines in the validation process. It will also help us understand the true impact of OS sub-versions in our ability to do provenance validation.
+
+### FAQ
+
+### What about binaries produced before VS 16.7 was released?
+The provenance validation process depends on extra metadata included in the PDB that the compiler uses to generate the artifacts manifest file as well as recreate the compilation. That extra information was first added into PDBs in the Visual Studio 16.7 release. That means any binaries produced before then will fail provenance validation. 
+
+This is a known issue and the understanding is such older NuPkgs will not be supported for provenance validation. 
+
+### How do we handle source generators?
+The provenance validation will occur using the output of source generators rather than re-running source generators as part of the validation. This will simplify the process as well as remove the issue of determinism in source generators (customer defined tools that don't have our level of rigor) from the process.
+
+## Open Issues
+
+### Running the validator on .NET Framework
+The expectation is that for Windows we should be able to use .NET Core to run the validator for binaries that were produced using the .NET Framework compiler. The runtime differences we are worried about for Unix operating systems don't have the same problems on Windows. Hence any .NET Runtime which can execute the compiler should be able to do provenance validation. 
+
+It's possible we will find differences here though in the real world. In which case we will need to have a mode for `dotnet-roslyn-validator` where by it will shell out to a .NET Framework process to do the validation. 
+
+### Line ending issues
+In order for validation to succeed it is important for `dotnet-roslyn-validator` to receive source files that have the same line endings as the one's provided to the original compilation. These factor into items like the checksum of source files stored in PDBs, the content of string literals, etc ... 
+
+It's possible that the information provided by source link is not sufficient to properly recreate files with correct line endings. Consider that a `.gitattributes` file could change how line endings are used on different operating systems. 
+
+We need to see how much of a problem this is in practice before we decide how to approach it. One possible solution is as simple as giving customers guidance on how to handle line endings in a provenance friendly manner.
+
+### Future expansions
+This document is written to describe how validation will work for the C# and VB compilers but those are not the only tools in the .NET SDK which create or modify DLLs. Other tools include, but is not limited to, the F# compiler and IL trimmer. 
```

Comment:
note that we run the IL trimmer on the Microsoft.NETCore.App shared framework as part of the dotnet/runtime build so that might be an issue if you want to support validating that one in v1.

Reply:
just want to confirm, so we're okay with Microsoft.NETCore.App to not be verifiable in v1?

Reply:
That is the current plan yes. The v1 focus is more on can we get provenance for packages outside the core framework, particularly those submitted by third parties. That in itself is going to be a large effort. Once we prove that out then we're going to step back and look towards getting other products like il trimmer involved.